airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/migrate_to_inline_schemas/pipeline.py

```diff
@@ -0,0 +1,381 @@
+
+from connector_ops.utils import ConnectorLanguage  # type: ignore
+from pipelines import main_logger
+from pipelines.airbyte_ci.connectors.consts import CONNECTOR_TEST_STEP_ID
+from pipelines.airbyte_ci.connectors.context import ConnectorContext, PipelineContext
+from pipelines.airbyte_ci.connectors.reports import Report
+from pipelines.consts import LOCAL_BUILD_PLATFORM
+from pipelines.helpers.connectors.command import run_connector_steps
+from pipelines.helpers.connectors.format import format_prettier
+from pipelines.helpers.connectors.yaml import read_yaml, write_yaml
+from pipelines.helpers.execution.run_steps import STEP_TREE, StepToRun
+from pipelines.models.steps import Step, StepResult, StepStatus
+
+if TYPE_CHECKING:
+    from anyio import Semaphore  # type: ignore
+
+SCHEMAS_DIR_NAME = "schemas"
+
+
+class CheckIsInlineCandidate(Step):
+    """Check if the connector is a candidate to get inline schemas.
+    Candidate conditions:
+    - The connector is a Python connector.
+    - The connector is a source connector.
+    - The connector has a manifest file.
+    - The connector has schemas directory.
+    """
+
+    context: ConnectorContext
+
+    title = "Check if the connector is a candidate for inline schema migration."  # type: ignore
+
+    def __init__(self, context: PipelineContext) -> None:
+        super().__init__(context)
+
+    async def _run(self) -> StepResult:  # type: ignore
+        connector = self.context.connector
+        manifest_path = connector.manifest_path
+        python_path = connector.python_source_dir_path
+        if connector.language not in [ConnectorLanguage.PYTHON, ConnectorLanguage.LOW_CODE]:
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector is not a Python connector.",
+            )
+        if connector.connector_type != "source":
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector is not a source connector.",
+            )
+
+        if not manifest_path.is_file():
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector does not have a manifest file.",
+            )
+
+        schemas_dir = python_path / SCHEMAS_DIR_NAME
+        if not schemas_dir.is_dir():
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector does not have a schemas directory.",
+            )
+
+        # TODO: does this help or not?
+        # if _has_subdirectory(schemas_dir):
+        #     return StepResult(step=self, status=StepStatus.SKIPPED, stderr="This has subdirectories. It's probably complicated.")
+
+        return StepResult(
+            step=self,
+            status=StepStatus.SUCCESS,
+        )
+
+
+def copy_directory(src: Path, dest: Path) -> None:
```

Comment:
Should this be in a util file somewhere? Should this be named explicit, i.e. `overwrite_copy_directory`?

Reply:
Honestly if you only use it twice, do you even have to DRY it?